spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DefaultDatastoreEntityConverter.java

```diff
@@ -93,27 +94,31 @@ public ReadWriteConversions getConversions() {
 		if (entity == null) {
 			return null;
 		}
-		DatastorePersistentEntity<R> persistentEntity = (DatastorePersistentEntity<R>) this.mappingContext
+		DatastorePersistentEntity<R> ostensiblePersistentEntity = (DatastorePersistentEntity<R>) this.mappingContext
 				.getPersistentEntity(aClass);
 
-		if (persistentEntity == null) {
+		if (ostensiblePersistentEntity == null) {
 			throw new DatastoreDataException("Unable to convert Datastore Entity to " + aClass);
 		}
 
 		EntityPropertyValueProvider propertyValueProvider = new EntityPropertyValueProvider(entity, this.conversions);
 
+		DatastorePersistentEntity persistentEntity = getDiscriminationPersistentEntity(ostensiblePersistentEntity,
+				propertyValueProvider);
+
 		ParameterValueProvider<DatastorePersistentProperty> parameterValueProvider =
 				new PersistentEntityParameterValueProvider<>(persistentEntity, propertyValueProvider, null);
 
 		EntityInstantiator instantiator = this.instantiators.getInstantiatorFor(persistentEntity);
-		R instance;
+		Object instance;
 		try {
 			instance = instantiator.createInstance(persistentEntity, parameterValueProvider);
 			PersistentPropertyAccessor accessor = persistentEntity.getPropertyAccessor(instance);
 			persistentEntity.doWithColumnBackedProperties((datastorePersistentProperty) -> {
 						// if a property is a constructor argument, it was already computed on instantiation
 						if (!persistentEntity.isConstructorArgument(datastorePersistentProperty)) {
-							Object value = propertyValueProvider.getPropertyValue(datastorePersistentProperty);
+					Object value = propertyValueProvider
+							.getPropertyValue((DatastorePersistentProperty) datastorePersistentProperty);
```

Comment:
if this cast necessary, can we just add type to lambda parameter?

Reply:
Done